### PR TITLE
Issue 203 auto switch cycle mode

### PR DESCRIPTION
**Pull request by yai-rosejo**
#203 
A continuation of the discussion and development of the cycle mode. These small changes allow Ros_MotionControl_StartMotionMode() to verify that the system is in a continuous cycle mode, and if it isn't it will try to set the cycle mode to continuous.

### CODE REVIEW

**Comment by ted-miller** [comment]
Just realized that I missed this. I'll validate in the next day or so. But the code changes look OK with a quick glance.

**Comment by gavanderhoorn** [comment]
I'm still wondering whether it would help raising an alarm if for some reason we can't change the cycle setting, in addition to the debug log statement.

(the service call will of course still fail with an appropriate message)

**Comment by ted-miller** [comment]
I agree that an alarm is warranted if the set-operation fails _and_ the current cycle is `STEP`.

**Review by ted-miller — CHANGES_REQUESTED** [review]
@yai-rosejo 
I just tested and it works great. 

Two things:
1) It would nice to implement that alarm.
2) I think it should do a final check to `Ros_Controller_IsContinuousCycle` to verify the setting was applied. This would require a call to `Ros_Controller_IoStatusUpdate` to fetch the value.

If you're tied up with Matt, let me know and I'll implement the changes.

**Review by ted-miller — APPROVED** [review]

File: src/MotionControl.c [hunk 1]
```diff
@@ -1436,6 +1430,34 @@ BOOL Ros_MotionControl_StartMotionMode(MOTION_MODE mode)
...
+        bzero(&sCycleData, sizeof(sCycleData));
+        bzero(&rData, sizeof(rData));
+        sCycleData.sCycle = 3;  // Auto
+        ret = mpSetCycle(&sCycleData, &rData);
+        if( (ret != 0) || (rData.err_no != 0) ) 
+        {
+            Ros_Debug_BroadcastMsg(
+                "Can't set cycle mode to continuous because: '%s' (0x%04X)",
+                Ros_ErrorHandling_ErrNo_ToString(rData.err_no), rData.err_no);
+            mpSetAlarm(ALARM_OPERATION_FAIL, "Set job-cycle to AUTO", SUBCODE_OPERATION_SET_CYCLE);
```
Comment (gavanderhoorn) [inline src/MotionControl.c:1447]: This message doesn't appear to correspond to what the documentation shows?

(same below)

Comment (ted-miller) [reply src/MotionControl.c:1447]: Oops. I changed my mind on the alarm msg after writing the doc. It is now updated.

File: doc/ros_api.md [hunk 2]
```diff
@@ -111,7 +111,7 @@ Note: errors and alarms which require physical operator intervention (e-stops, e
 
 Type: [motoros2_interfaces/srv/StartTrajMode](https://github.com/yaskawa-global/motoros2_interfaces/blob/d6805d32714df4430f7db3d8ddc736c340ddeba8/srv/StartTrajMode.srv)
 
-Attempt to enable servo drives and activate trajectory mode, allowing the action server (`follow_joint_trajectory`, see below) to execute incoming `FollowJointTrajectory` action goals.
+Attempts to set cycle mode, enable servo drives, and activate trajectory mode, allowing the action server (`follow_joint_trajectory`, see below) to execute incoming `FollowJointTrajectory` action goals.
```
Comment (gavanderhoorn) [inline doc/ros_api.md:114]: we're probably not going to do anything about it, but it's a bit 'mysterious' to have this state it's going to *set cycle mode* without any motivation. Enabling the servos and activating trajectory mode all makes sense, but changing cycle mode seems odd.

File: doc/troubleshooting.md [hunk 3]
```diff
@@ -1117,3 +1117,24 @@ ALARM 8015
...
+*Example:*
+
+```text
+ALARM 8016
+ Set job-cycle to AUTO
+[0]
+```
+
+*Solution:*
+The job cycle is currently set to `STEP` and MotoROS2 was unable to automatically change it to `AUTO`.
```
Comment (gavanderhoorn) [inline doc/troubleshooting.md:1132]: should we perhaps add some clarification as to *why* this is needed?

File: doc/troubleshooting.md [hunk 4]
```diff
@@ -1117,3 +1117,24 @@ ALARM 8015
...
+
+First, manually set the controller to use `AUTO`.
```
Comment (gavanderhoorn) [inline doc/troubleshooting.md:1134]: `AUTO` what?

File: doc/troubleshooting.md [hunk 5]
```diff
@@ -1117,3 +1117,24 @@ ALARM 8015
...
+
+ 1. upgrade to *MANAGEMENT* security level by touching `[System Info]`→`[Security]` (default password is all `9`'s)
+ 1. touch `[Setup]`→`[Operate Cond.]`
+ 1. change `CYCLE SWITCH IN REMOTE MODE` to `AUTO`
+
+If the problem persists, verify that the `CIOPRG.LST` ladder program is not writing to `#40050 - #40052`.
```
Comment (gavanderhoorn) [inline doc/troubleshooting.md:1140]: I expect we'll receive support requests here, because this is rather cryptic.

File: src/MotionControl.c [hunk 6]
```diff
@@ -1436,6 +1430,34 @@ BOOL Ros_MotionControl_StartMotionMode(MOTION_MODE mode)
...
     }
 
+    // Check if in continous cycle mode
+    if (!Ros_Controller_IsContinuousCycle())
+    {
+        // set the cycle mode to auto if not currently
+        MP_CYCLE_SEND_DATA sCycleData;
+        bzero(&sCycleData, sizeof(sCycleData));
+        bzero(&rData, sizeof(rData));
+        sCycleData.sCycle = 3;  // Auto
```
Comment (gavanderhoorn) [inline src/MotionControl.c:1440]: M+ does not define any named constant(s) for this?

File: src/MotionControl.c [hunk 7]
```diff
@@ -1436,6 +1430,34 @@ BOOL Ros_MotionControl_StartMotionMode(MOTION_MODE mode)
...
+        ret = mpSetCycle(&sCycleData, &rData);
+        if( (ret != 0) || (rData.err_no != 0) ) 
+        {
+            Ros_Debug_BroadcastMsg(
+                "Can't set cycle mode to continuous because: '%s' (0x%04X)",
+                Ros_ErrorHandling_ErrNo_ToString(rData.err_no), rData.err_no);
```
Comment (gavanderhoorn) [inline src/MotionControl.c:1446]: is `rData.err_no` guaranteed to be non-zero in case `ret != 0`?

File: src/MotionControl.c [hunk 8]
```diff
@@ -1436,6 +1430,34 @@ BOOL Ros_MotionControl_StartMotionMode(MOTION_MODE mode)
...
+            mpSetAlarm(ALARM_OPERATION_FAIL, "Set job-cycle to AUTO", SUBCODE_OPERATION_SET_CYCLE);
+            return FALSE;
+        }
+
+        Ros_Controller_IoStatusUpdate(); //verify the cycle got set and wasn't forced back due to CIO logic
```
Comment (gavanderhoorn) [inline src/MotionControl.c:1451]: would this require a delay first?

**Comment by gavanderhoorn** [comment]
Friendly ping?

**Comment by ted-miller** [comment]
Apologies for the delay. I will look at this again later in the week.

File: doc/ros_api.md (hunk 2 above)
Comment (ted-miller) [reply doc/ros_api.md:114]: Are you saying that you'd prefer we remove this text? 

I think it's implied that if this service is performing a particular action, then there's a good reason for it.

File: src/MotionControl.c (hunk 6 above)
Comment (ted-miller) [reply src/MotionControl.c]: It does not

File: src/MotionControl.c (hunk 7 above)
Comment (ted-miller) [reply src/MotionControl.c:1446]: Yes

File: src/MotionControl.c (hunk 8 above)
Comment (ted-miller) [reply src/MotionControl.c:1452]: Yeah. good catch

File: doc/troubleshooting.md (hunk 5 above)
Comment (ted-miller) [reply doc/troubleshooting.md:1141]: I added text to contact yaskawa directly

File: doc/ros_api.md (hunk 2 above)
Comment (gavanderhoorn) [reply doc/ros_api.md:114]: > Are you saying that you'd prefer we remove this text?

No, just that the part about this 'attempting to set cycle mode' is a bit cryptic. The other actions it performs 'make sense': can't move the robot without the servos being enabled or trajectory mode being activated.

But then it also *sets cycle mode*, which to me at least, is far from clear why.

But perhaps this is too much of a nit.

> I think it's implied that if this service is performing a particular action, then there's a good reason for it.

Of course. I just always like to understand *why* something happens.

**Comment by gavanderhoorn** [comment]
I've added (a) named constant(s) in a461aeed7ecf9f5daf9e1806c05241ac5c4e4e0f and rebased on top of `main` as this was otherwise in conflict (due to move of header to the new `lib` sub dir).

**Comment by ted-miller** [comment]
build-check was verified locally.

**Pull request merged**